pkg/descheduler/strategies/topologyspreadyconstraint.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategies
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"github.com/kubernetes-incubator/descheduler/cmd/descheduler/app/options"
+	"github.com/kubernetes-incubator/descheduler/pkg/api"
+	"github.com/kubernetes-incubator/descheduler/pkg/descheduler/evictions"
+	podutil "github.com/kubernetes-incubator/descheduler/pkg/descheduler/pod"
+)
+
+func TopologySpreadConstraint(ds *options.DeschedulerServer, strategy api.DeschedulerStrategy, policyGroupVersion string, nodes []*v1.Node, nodePodCount nodePodEvictedCount) {
+	if !strategy.Enabled {
+		return
+	}
+
+	fmt.Printf("Found following parameters for TopologySpreadConstraint %v\n", strategy)
+	for _, topoConstraints := range strategy.Params.NamespacedTopologySpreadConstraints {
+		glog.Infof("Found Topo Constraints %v\n", topoConstraints)
+		evictPodsViolatingSpreadConstraints(ds.Client, policyGroupVersion, nodes, ds.DryRun, nodePodCount, topoConstraints)
+	}
+
+}
+
+// AntiAffinityTerm's topology key value used in predicate metadata
+type topologyPair struct {
+	key   string
+	value string
+}
+
+type podSet map[*v1.Pod]struct{}
+
+type topologyPairSet map[topologyPair]struct{}
+
+// finnd all nodes
+// find all pods on the nodes
+// find the pods that match the namespace
+// find the constraints topokey , use this topokey to finds the node topokey value
+// add the pod with key as the combination of topokey and its value
+// find the min value of pods with any key
+// iterate through all keys and diff currentPods -minPods <=maxSkew
+// if diff > maxSkew, select a pod in the current bucket for eviction
+//
+func evictPodsViolatingSpreadConstraints(client clientset.Interface, policyGroupVersion string, nodes []*v1.Node, dryRun bool, nodePodCount nodePodEvictedCount,
+	constraint api.NamespacedTopologySpreadConstraint) {
+
+	if len(constraint.TopologySpreadConstraints) != 1 {
+		glog.V(1).Infof("We currently only support 1 topology spread constraint per namespace")
```

Comment:
I think, we need to explicitly document this along with reason.

Reply:
Agree. This is a significant limitation.

Reply:
I'm lean to delivering a full implementation which respects all constraints.

Reply:
@Huang-Wei so per namespace, do an AND of all constraints ?

Reply:
@krmayankk Yes.